code/go/internal/validator/folder_spec.go

```diff
@@ -119,127 +91,94 @@ func (s *folderSpec) validate(pkg *Package, folderPath string) ve.ValidationErro
+				v.totalContents += itemValidator.totalContents
+				v.totalSize += itemValidator.totalSize
 			}
 		} else {
-			if !itemSpec.isSameType(file) {
+			if itemSpec.IsDir() {
 				errs = append(errs, fmt.Errorf("[%s] is a file but is expected to be a folder", pkg.Path(fileName)))
 				continue
 			}
 
 			itemPath := path.Join(folderPath, file.Name())
-			itemValidationErrs := itemSpec.validate(s.fs, pkg, s.specPath, itemPath)
-			if itemValidationErrs != nil {
-				for _, ive := range itemValidationErrs {
-					errs = append(errs, errors.Wrapf(ive, "file \"%s\" is invalid", pkg.Path(itemPath)))
-				}
+			itemValidationErrs := validateFile(itemSpec, pkg, itemPath)
+			for _, ive := range itemValidationErrs {
+				errs = append(errs, errors.Wrapf(ive, "file \"%s\" is invalid", pkg.Path(itemPath)))
 			}
 
 			info, err := fs.Stat(pkg, itemPath)
 			if err != nil {
 				errs = append(errs, errors.Wrapf(err, "failed to obtain file size for \"%s\"", pkg.Path(itemPath)))
 			} else {
-				s.totalContents++
-				s.totalSize += spectypes.FileSize(info.Size())
+				v.totalContents++
+				v.totalSize += spectypes.FileSize(info.Size())
 			}
 		}
 	}
 
-	if sizeLimit := s.Limits.TotalSizeLimit; sizeLimit > 0 && s.totalSize > sizeLimit {
+	if sizeLimit := spec.MaxTotalSize(); sizeLimit > 0 && v.totalSize > sizeLimit {
 		errs = append(errs, errors.Errorf("folder [%s] exceeds the total size limit of %s", pkg.Path(folderPath), sizeLimit))
 	}
 
 	// validate that required items in spec are all accounted for
-	for _, itemSpec := range s.Contents {
-		if !itemSpec.Required {
+	for _, itemSpec := range spec.Contents() {
+		if !itemSpec.Required() {
 			continue
 		}
 
-		fileFound, err := itemSpec.matchingFileExists(files)
+		fileFound, err := matchingFileExists(itemSpec, files)
 		if err != nil {
 			errs = append(errs, err)
 			continue
 		}
 
 		if !fileFound {
 			var err error
-			if itemSpec.Name != "" {
-				err = fmt.Errorf("expecting to find [%s] %s in folder [%s]", itemSpec.Name, itemSpec.ItemType, pkg.Path(folderPath))
-			} else if itemSpec.Pattern != "" {
-				err = fmt.Errorf("expecting to find %s matching pattern [%s] in folder [%s]", itemSpec.ItemType, itemSpec.Pattern, pkg.Path(folderPath))
+			if itemSpec.Name() != "" {
+				err = fmt.Errorf("expecting to find [%s] %s in folder [%s]", itemSpec.Name(), itemSpec.Type(), pkg.Path(folderPath))
+			} else if itemSpec.Pattern() != "" {
+				err = fmt.Errorf("expecting to find %s matching pattern [%s] in folder [%s]", itemSpec.Type(), itemSpec.Pattern(), pkg.Path(folderPath))
 			}
 			errs = append(errs, err)
 		}
 	}
 	return errs
 }
 
-func (s *folderSpec) findItemSpec(packageName string, folderItemName string) (*folderItemSpec, error) {
-	for _, itemSpec := range s.Contents {
-		if itemSpec.Name != "" && itemSpec.Name == folderItemName {
-			return &itemSpec, nil
+func findItemSpec(s spectypes.ItemSpec, packageName string, folderItemName string) (spectypes.ItemSpec, error) {
```

Comment:
Should this function be defined as part of the `validator struct` ? As it is `validateSpec()`, something like:
```
func (v *validator) findItemSpec(s spectypes.ItemSpec, packageName string, folderItemName string) (spectypes.ItemSpec, error) {
```
If so, probably those parameters could be accessed from the validator `v` directly.

Reply:
Makes sense, yes :+1:

Reply:
Done.